### PR TITLE
OPQ rotation matrix training

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/src/index/vector/opq.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            let reconstructed_vector = pq.reconstruct(code.values());
+            builder.append_slice(reconstructed_vector.values());
+        }
+        // Reconstructed vectors, `Y` in Section 3.1 in CVPR' 13.
+        // X and Y are column major described in the paper.
+        let y_data = Arc::new(builder.finish());
+        let y = MatrixView::new(y_data, dim).transpose();
+        // Solving `min||RX - Y||` in Section 3.1 in CVPR' 13.
+        //
+        //  X * T(Y) = U * S * T(V)
```
Comment (changhiskhan) [inline rust/src/index/vector/opq.rs:124]: isn't this missing the reordering by eigenvalues?

File: rust/src/arrow/linalg.rs [hunk 2]
```diff
@@ -75,6 +74,22 @@ impl MatrixView {
         }
     }
 
+    /// Randomly initialize a matrix of shape `(num_rows, num_columns)`.
+    pub fn random(num_rows: usize, num_columns: usize) -> Self {
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply rust/src/arrow/linalg.rs:78]: https://numpy.org/doc/stable/reference/random/generated/numpy.random.rand.html This is basically doing `numpy.random.rand(shape)` for 2D matrix ?

[the remark this answers is not in the document]
Comment (changhiskhan) [reply rust/src/arrow/linalg.rs:78]: Oh you are. First glance I thought this was taking random indices from existing matrix

**Review by changhiskhan — APPROVED** [review]
Pending unit test for QR

**Pull request merged**